geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyFragment.kt

```diff
@@ -0,0 +1,631 @@
+import org.odk.collect.maps.layers.OfflineMapLayersPickerBottomSheetDialogFragment
+import org.odk.collect.maps.layers.ReferenceLayerRepository
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.webpage.WebPageService
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+class GeoPolyFragment @JvmOverloads constructor(
+    val outputMode: OutputMode = OutputMode.GEOTRACE,
+    val readOnly: Boolean = false,
+    val retainMockAccuracy: Boolean = false,
+    val inputPolygon: List<MapPoint> = emptyList(),
+    val invalidMessage: LiveData<String?> = MutableLiveData(null)
+) : Fragment(R.layout.geopoly_layout), SettingsDialogCallback {
+
+    @Inject
+    lateinit var mapFragmentFactory: MapFragmentFactory
+
+    @Inject
+    lateinit var locationTracker: LocationTracker
+
+    @Inject
+    lateinit var referenceLayerRepository: ReferenceLayerRepository
+
+    @Inject
+    lateinit var scheduler: Scheduler
+
+    @Inject
+    lateinit var settingsProvider: SettingsProvider
+
+    @Inject
+    lateinit var webPageService: WebPageService
+
+    private var previousState: Bundle? = null
+    private val executorServiceScheduler: ScheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor()
+    private var schedulerHandler: ScheduledFuture<*>? = null
+
+    private var map: MapFragment? = null
+    private var featureId = -1 // will be a positive featureId once map is ready
+    private var originalPoly: List<MapPoint>? = null
+
+    private var zoomButton: ImageButton? = null
+    var playButton: ImageButton? = null
+    var clearButton: ImageButton? = null
+    private var recordButton: Button? = null
+    private var pauseButton: ImageButton? = null
+    var backspaceButton: ImageButton? = null
+    var saveButton: ImageButton? = null
+
+    private var locationStatus: AccuracyStatusView? = null
+    private var collectionStatus: TextView? = null
+
+    private var settingsView: View? = null
+
+    private var inputActive = false // whether we are ready for the user to add points
+    private var recordingEnabled =
+        false // whether points are taken from GPS readings (if not, placed by tapping)
+    private var recordingAutomatic =
+        false // whether GPS readings are taken at regular intervals (if not, only when user-directed)
+
+    private var intervalIndex: Int = DEFAULT_INTERVAL_INDEX
+
+    private var accuracyThresholdIndex: Int = DEFAULT_ACCURACY_THRESHOLD_INDEX
+
+    // restored from savedInstanceState
+    private var restoredPoints: List<MapPoint>? = null
+
+    private val onBackPressedCallback: OnBackPressedCallback =
+        object : OnBackPressedCallback(true) {
+            override fun handleOnBackPressed() {
+                if (!readOnly && map != null && (originalPoly != map!!.getPolyLinePoints(featureId))) {
```

Comment:
The parentheses around `originalPoly != map!!.getPolyLinePoints(featureId)` are not required